### DISCUSSION TITLE
The best way to use the Sanity Client from within the new Studio

### DISCUSSION THREAD

**Discussion by franknoirot** [sanity-io/sanity#3359] [Sanity Studio Support] [answered]
I was wondering what the path for using the client from within the studio was when the part system is gone? 

Under V2 whenever I needed to query some content for use in odd places where no client was passed into the context I needed it, I would import from the parts system like so:
```js
import sanityClient from 'part:@sanity/base/client'

const client = sanityClient.withConfig({ apiVersion: '2022-06-17' })
```

Now there is a new `useClient` React hook, but I'm not sure the best way to use the client within the Studio when I don't have a React component. My specific example is as follows: I am making an Initial Value Template for creating a Product entry based on a given Product Category's ID field value. I want to scaffold out some data based on content within the Category, but I need to fetch the document based on that ID in order to do so. The TODO comments below show where I'd like to fetch and make use of a document with the client.
```js
// sanity.config.js

export default createConfig({
  // ...more configuration above
  document: {
      newDocumentOptions: (prev, {currentUser, creationContext}) => {
        const verticalCategoryId = 'cebd3e27-22cc-44db-9603-5fdaf1695f8d'
  
        // TODO get the product options defined within the category to use below
  
        return [
          ...prev,
          {
            id: 'productByCategorySingle',
            templateId: 'product-by-category',
            type: 'initialValueTemplateItem',
            title: 'Vertical Photography',
            parameters: {
              categoryId: verticalCategoryId,
              variants: [] // TODO: scaffold out the variants based on product options 
            },
          },
        ]
      }
    }
}
```

I could definitely import the client the same way I would from a frontend application, but I figure I should check best practices since I'm already in the Studio context. Thanks!

**Comment by geball** [comment]
Hi Frank. It looks like the v3 team has [recommended](https://github.com/sanity-io/sanity/discussions/3329#discussioncomment-2955321) using dependency injection, instantiating the store in a `useMemo` with the client as a dependency. However, they [also mentioned](https://github.com/sanity-io/sanity/discussions/3329#discussioncomment-2962131) that they're considering a way to wrap the entire studio, so that recommendation above may eventually change.

**Reply by franknoirot** [reply]
Thanks for the quick reply @geball! I might not understand correctly, but I don't think the `useMemo` solution will work in my case, as I don't think the callback that can be passed into `newDocumentOptions` is a React component, and may only be either an array of template objects or a callback [based on the docs](https://beta.sanity.io/docs/platform/studio/v2-to-v3/new-document-structure) (sidenote: I don't know if it would support a Promise that resolves to an array of template objects).

The V2 statement `import sanityClient from 'part:@sanity/base/client'` could be used in plain JavaScript callbacks like the one I'm passing to `newDocumentOptions`. I guess what I was expecting was for there to be a plain JS client available somewhere within the callback, such as a part of that second parameter that passes in `currentUser` and `creationContext`:
```js
// ...
newDocumentOptions: async (prev, { currentUser, creationContext, client }) => {
  const verticalCategoryId = 'cebd3e27-22cc-44db-9603-5fdaf1695f8d'
  
  // TODONE get the product options defined within the category to use below
  const categoryOptions = await client.getDocument(verticalCategoryId)?.options

// ...
```

**Reply by coreyward** [reply]
+1 on the need for a way to do this outside/independent of React components. I most often reach for the client “part” for validations and other schema-related callbacks (e.g. readonly and hidden) and for configuring the desk structure. Without a low-friction, official way of doing this, I suspect many will end up hardcoding credentials.

**Reply by jfulse** [reply]
Hello! Have you found a canonical way of accessing the client within the schema configuration?

**Reply by franknoirot** [reply]
I haven't so far @jfulse, but I'll revisit the docs and my prototype I was hacking on to see if anything has changed. If all else fails I believe you can use the Sanity client as you would normally from outside your studio, with the minor annoyance of having to configure it and have it running in addition to the Sanity Studio itself.

**Reply by coreyward** [reply]
As a hack, you might try setting `window.currentUser = useCurrentUser()` in a React component mounted somewhere in the tree. Hopefully this comment becomes obsolete quickly. 🤞

**Reply by rexxars** [reply]
The `newDocumentOptions` should receive both `client` and `currentUser` within the context, just as you suggested @franknoirot  - however the resolver cannot currently return an asynchronous value. That is an interesting use case however, so we'll look into if we can add support for that.

**Reply by 16n** [reply]
Is there any news regarding this? I have a use case where I would like to prefill a set of references which should be fetched async with the sanity client.

**Reply by rexxars** [reply]
Depends where/how you want to do it. Are you wanting to do it from initial values? If so, `initialValue` can be defined as a function - the second argument is a context that contains a `getClient()` method:

```js
{
  name: 'categories',
  type: 'array',
  of: [{type: 'reference', to: [{type: 'category'}]}],
  initialValue: (params, context) => {
    const client = context.getClient({apiVersion: '2022-11-01'})
    return client.fetch('*[_type == "category"]{_ref: _id, _key: _id}')
  }
}
```

If you're doing it from a custom input component or other React component, you can use the `useClient` hook:

```js
import {useClient} from 'sanity'

function MyComponent() {
  const client = useClient({apiVersion: '2022-11-01'})
  // do something with the client
}
```

**Reply by 16n** [reply]
thank you very much - I did not know that I can return a Promise for the initialValue. This is exactly what I was looking for.

**Reply by franknoirot** [reply]
> The `newDocumentOptions` should receive both `client` and `currentUser` within the context, just as you suggested @franknoirot - however the resolver cannot currently return an asynchronous value. That is an interesting use case however, so we'll look into if we can add support for that.

@rexxars am I understanding the most recent [v3.7.0 release notes](https://github.com/sanity-io/sanity/releases/tag/v3.7.0) correctly that you've solved my original question with the new `getClient()` hook? Or is not being able to return asynchronous values from that resolver still the main limiting factor?

**Reply by rexxars** [reply, marked as answer]
Unfortunately the new document options resolver is still synchronous - this release only adds a `getClient()` _function_ (not a React hook) to a specific context - the one passed to reference resolvers.
